certloader/acmetlsconfig.go

```diff
@@ -0,0 +1,86 @@
+package certloader
+
+import (
+	"crypto/tls"
+	"errors"
+
+	"github.com/caddyserver/certmagic"
+	"github.com/mholt/acmez"
+)
+
+func TLSConfigSourceFromACME(domain, email string, useTestCA bool) (TLSConfigSource, error) {
+	certmagic.DefaultACME.Agreed = true
+	certmagic.DefaultACME.DisableHTTPChallenge = true
+	certmagic.DefaultACME.Email = email
+
+	if useTestCA == true {
+		certmagic.DefaultACME.CA = certmagic.LetsEncryptStagingCA
+	}
+
+	magicConfig := certmagic.NewDefault()
+
+	// Force an iniial synchronous load of the certificate on startup. If no certificate
+	// yet exists, certmagic will attempt to obtain one from the ACME provider. If a valid
+	// cert has already been obtained, it will be loaded from local cache.
+	err := magicConfig.ManageSync([]string{domain})
+	if err != nil {
+		return nil, err
+	}
+
+	return &acmeTLSConfigSource{
+		magicConfig: magicConfig,
+		domain:      domain,
+		email:       email,
+	}, nil
+}
+
+type acmeTLSConfigSource struct {
+	magicConfig *certmagic.Config
+	domain      string
+	email       string
+}
+
+func (a *acmeTLSConfigSource) Reload() error {
+	// certmagic automatically keeps certs updated
+	return nil
+}
+
+func (a *acmeTLSConfigSource) CanServe() bool {
+	// It does not appear that certmagic currently provides any API for determining
+	// if it has a valid cert/key/chain for serving TLS connections. For now, we
+	// assume certmagic correctly does its thing.
```

Comment:
I agree this is a safe assumption. `CanServe` should only return false in a client mode where no client certificate exists, since ACME is only supported in server mode at the moment it's safe to always return `true`.